### DISCUSSION TITLE
Get an exception outside of Marten during a projection

### DISCUSSION THREAD

**Discussion by NicolasREY69330** [JasperFx/marten#2375] [General]
Hi,

following the advices from @oskardudycz I implemented the `IProjection`  interface allowing me to have a custom behavior around the domain events stored into Postgres (event sourcing).

**I wonder if it's possible to get an exception thrown into the projection background worker ?**

For example I intentionally made the deserialization fail (deleting/renaming the class of a domain event already serialized and stored in Postgres) to make the deserialization fail. I see Marten logging the `EventFetcherException`, but I would like to be given an access to it, adding some custom behavior (tracing, pushing external events and so on ...).

For example, adding the following method to the IProjection interface + ad hoc implementation

EDITED on 02/11/22 (I misuse my internal IProjection interface instead of Marten's one ...)
``` csharp
    public interface IProjection
    {
          Task ExceptionOnApplyAsync(Exception exception, IDocumentOperations operations, IReadOnlyList<StreamAction> streams, CancellationToken cancellation); // <== new method triggered on exception
    }
```
Thank you very much for this awesome lib, my company has chosen it to implement ES in production.

**Comment by jeremydmiller** [comment]
It's absolutely possible to get exceptions in the async projections. The daemon has configurable exception handling to do limited retries, or cooldown, or skip poison pill events

**Comment by NicolasREY69330** [comment]
Currently here is my projection

``` csharp
    public class MartenAutoProjectionManager : IProjection
    {
        private readonly IServiceProvider _serviceProvider;
        private readonly ILogger<MartenAutoProjectionManager> _logger;

        public MartenAutoProjectionManager(IServiceProvider serviceProvider)
        {
            _serviceProvider = serviceProvider;
            _logger = _serviceProvider.GetService<ILogger<MartenAutoProjectionManager>>();
        }

        public void Apply(IDocumentOperations operations, IReadOnlyList<StreamAction> streams)
        {
            throw new NotImplementedException();
        }

        public async Task ApplyAsync(IDocumentOperations operations, IReadOnlyList<StreamAction> streams, CancellationToken cancellation)
        {
            var events = streams.SelectMany(x => x.Events).OrderBy(s => s.Sequence).Select(s => s);
            var projections = _serviceProvider.GetServices<Projections.Interfaces.IProjection>();
            foreach (var @event in events)
            {
                   await Task.WhenAll(projections.Select(x => x.Project(@event.Data, @event.Id.ToString())));
            }
        }
    }
```

I would like to be able to get triggered with the exception thrown + the `IDocumentOperations operations`, `IReadOnlyList<StreamAction> streams` objects to decide what to do.

I've seen error handling documentation https://martendb.io/events/projections/async-daemon.html#error-handling
But unless I'm wrong, I can't access to the event responsible for the exception itself here ?

Thank you for your help
